### PR TITLE
Add support for immutable_parameters in action aliases

### LINKED ISSUE #1704: Immutable/default action parameters for aliases (ChatOps)
There is already mechanism to set [optional parameter in alias](http://docs.stackstorm.com/latest/aliases.html#with-default):

``` yaml
formats:
  - "google {{query=StackStorm}}"
```
#### Problem

But what if we want to add simple hardcoded chatops command and hide unneeded logic behind the default/immutable action parameters?
I think that good ChatOps commands could be simple and easy to remember, without MANY optional parameters. Example case: 
deploy chatops

To do that, we need to create additional action with default parameters, which will trigger another action. I do this way almost everywhere (bad way): https://github.com/armab/st2-chatops-aliases/tree/master/actions
#### Proposal

The proposal is to allow setting from alias default/immutable action parameters.

Here is an example how it could look:

``` yaml

---
name: chatops.deploy
action_ref: packs.install
description: Deploy ChatOps pack from `armab/st2-ansible-chatops` GitHub repo
formats:
  - "deploy chatops"
# these lines were added
# naming/structure is doubtful, but you got the idea
action_parameters:
  packs: "st2-ansible-chatops"
  repo_url: "armab/st2-ansible-chatops"
```

Here you can see that behind simple and immutable/hardcoded ChatOps command:

```
!deploy chatops
```

stands: 

``` sh
st2 run packs.install packs=st2-ansible-chatops repo_url=armab/st2-ansible-chatops
```

---

This simplifies things a lot, making existing actions more reusable by aliases.

### PR DESCRIPTION
**Pull request by nicholasamorim**
Hi, this is an attempt to solve #1704.

Closes #1704 

There's an use case that consistently comes up where a dev wants a certain action alias but keep certain action parameters static. The current workarounds are to either have some really complex regexes or to create redundant dummy actions.

This proposal solves this by adding `action_parameters` parameter to the action alias definition. Here's a very simple example:

```yaml
---
name: "remote_static_shell_cmd"
action_ref: "core.remote"
description: "Execute a command the development machine"
formats:
  - "run {{cmd}} on dev"
action_parameters:
  hosts: the-dev-server
```

Whatever is defined in `action_parameters` will be passed to the action unmodified. This is a very simple PR and I'd love to get suggestions on how to improve the concept and/or the tests.

### CODE REVIEW

**Comment by nmaludy** [comment]
One small feature enhancement would be to evaluate Jinja expressions for the static variables, that way we can load from the datastore.

**Comment by armab** [comment]
> One minor complaint with this approach is that st2chatops does not display the static parameters with the !help command:

I think this is good that chatops doesn't display static parameters and hiding defaults. That's one of the simplicity/abstraction advantages behind this feature. 
The same could be achieved before by someone creating intermediate action with setting action param values explicitly.
There is also a way to control this how the `!help` shows command with `formats.display`.

----

Priority/overriding params is a good catch.
Setting and overriding default parameters in alias could be already achieved by `formats.representation` jinja/regex hackery as described in https://github.com/StackStorm/st2/issues/1704#issuecomment-179614192 and https://github.com/StackStorm/st2/issues/1704#issuecomment-180020375

For this feature I'd stick to initial https://github.com/StackStorm/st2/issues/1704 with _immutable_ action params and what @blag suggests, meaning something that's defined in alias can't be changed by the chat user.

We need to make sure we test these edge cases.

If we go that path, I also think it may be more clear to stick with `immutable_parameters` naming.
Immutability comes from the fact that param explicitly configured by the action alias creator. Dynamic nature in case of datastore used wouldn't break it.

**Comment by blag** [comment]
> I think that a more suitable approach would be to simply invert the logic: parameters passed on the format overrides the defaults, very much like one would expect. What do you think, @blag?

I'm not sure I understand this completely. How is this any different than our [(mutable) optional variables with default values](https://docs.stackstorm.com/chatops/aliases.html#with-default)?

I agree with @armab - the usefulness of explicitly specifying action parameters in the alias is that they are immutable to ChatOps users, so the `action_parameters` or `immutable_parameters` specified in the action alias should override any parameters specified by the ChatOps user, or it should at least cause an error if the ChatOps user tries to override the immutable parameters by specifying them explicitly.

And since immutability is the important part of this feature (IMO), I think using the `immutable_parameters` key is a little bit more clear than `action_parameters`.

> Not sure if I follow, you mean different static parameters for different formats? If this is what you mean, I absolutely agree: a different set of static parameters is a new action alias.

Sorry if I wasn't clear - that is exactly what I mean.

TL;DR: Not showing the immutable action parameters in the `!help` response and silently overriding values specified by the user is fine.

We have four different ways we can implement this (the 🛑  and ✅ are my opinion):

|                  |Warn/error on user-provided immutable parameters | Silently override user-provided parameters with their immutable values |
| ---------- | ---------------------------------------------------- | ----------------------------------------------------------------------- |
| Show immutable parameters in `!help` |        1. ✅            |            2. 🛑              |
| Hide immutable parameters in `!help` |          3. 🛑            |            4. ✅              |

1. If we show the immutable parameters in `!help`, it makes sense to me to error out or at/ least warn the user that whatever values they specify for immutable parameters will be overridden.
2. Showing immutable variables and silently overriding them makes it too easy to write aliases that can cause issues.
3. It doesn't make sense to me to hide the immutable parameters in `!help` and then warn/error if a user specifies them, because then there is relevant information that isn't provided in `!help` responses.
4. However, hiding the immutable parameters in `!help` and silently overriding them if the user does specify them is fine, because if the immutable parameters aren't ever shown to the user in the `!help` response, ChatOps users shouldn't be specifying them anyway.

Aliases are kind of the ChatOps "interface" to actions, which are kind of the "implementation". And I guess hiding implementation details is fine as long as the interface (and documentation via the `!help` response) is "good enough".

**Comment by nicholasamorim** [comment]
> I'm not sure I understand this completely. How is this any different than our (mutable) optional variables with default values?

It's not different indeed in the effective outcome but leads to more natural format messages while still leaving room for overriding those defined parameters. 

I mean, I'm throwing feedback on the opinions put here and try to conciliate the views and understand where I should lead the PR.

My _personal_ opinion is:

- The immutable parameters _cannot_ be overriden via the chat message/format.
- The immutable parameters, though, can be _dynamically loaded/evaluated_ using Jinja.
- If a user happens to try and override it, I would throw a hard error: you cannot override it and if you wish so, then either:
  a. Create another alias
  b. Remove it from the immutable parameters and have it on the `format`.
- On naming: `action_parameters` would be my (personal) preference but `immutable_parameters` makes sense as well and is indeed clearer when _associated with_ aliases, i.e: "immutable in the context of this alias".
---

Regarding showing static parameters on `!help`:

I think we should definitely show it. In fact, I don't even think there's a good reason to _not show_ it. It is part of the alias and often cases, it will be a _fundamental_ part of the alias as I'm sure a lot of users will use that to, for example, define an environment, e.g: "this can only happen in `staging`", "this will only run in `prod`" and so forth. This is a simplistic use case but the static parameters have direct impact on the decision an user might make.

**Comment by armab** [comment]
Cool, that sounds like a more or less consensus :+1:

**Comment by Kami** [comment]
I'm personally +1 for a change like that since I believe it makes it more clear and obvious what is going on and also solves the immutable default parameters use case (which doesn't require user to write a custom wrapper action to handle that and specify immutable default parameter values).

I also believe that at some point we also supported some syntax similar to this one, before the large action alias refactor a couple of years ago.

**Comment by nicholasamorim** [comment]
- I've renamed the key to be `immutable_parameters`.
- `ValueError` is now raised if the user tries to override them.
- Jinja expressions are now evaluated.

This last point, I'm not sure if I did the right thing.I couldn't find some function that would already give me the Datastore context and so on so I basically copied the `_get_rendered_vars` method from the `ChainHolder` class (`action_chain_runnerpy`). Is there a better way?

File: st2tests/st2tests/fixtures/aliases/aliases/alias5.yaml [hunk 1]
```diff
@@ -0,0 +1,10 @@
+---
+    name: "alias5"
+    pack: "aliases"
+    description: "Static test"
+    action_ref: "wolfpack.action1"
+    formats:
+        - "lorem ipsum"
+    static:
```
[the remark this answers is not in the document]
Comment (nicholasamorim) [reply st2tests/st2tests/fixtures/aliases/aliases/alias5.yaml]: Renamed to `immutable_parameters`

**Comment by nicholasamorim** [comment]
I have also opened a `st2docs` PR regarding this feature: https://github.com/StackStorm/st2docs/pull/933

File: CHANGELOG.rst [hunk 2]
```diff
@@ -10,7 +10,7 @@ Added
 * Add support for blacklisting / whitelisting hosts to the HTTP runner by adding new
   ``url_hosts_blacklist`` and ``url_hosts_whitelist`` runner attribute. (new feature)
   #4757
-* Add ``user`` parameter to ``re_run`` method of st2client. #4785
```
[the remark this answers is not in the document]
Comment (nicholasamorim) [reply CHANGELOG.rst]: Yes, that was a mistake, sorry. I'll fix it.

[the remark this answers is not in the document]
Comment (nicholasamorim) [reply CHANGELOG.rst]: Fixed.

**Comment by blag** [comment]
I'll double up on that gratitude. This is something our users have been requesting for awhile.
